### ISSUE TITLE
About the inference speedup

### ISSUE THREAD

**Issue by yimingsh** [ChandlerGuan/Transkimmer#2]
Hi~ @ChandlerGuan An interesting work.
I test the Transkimmer in the inference stage compared with BERT using the torchprofile repo. I found the MAC of Transkimmer is even larger than the original BERT.
Then I noticed that in your code, the model still uses the whole length of input in the evaluation stage, but actually, the length is reduced from the bottom layer to the top layer. Therefore, how did you evaluate the inference speedup in Table3 of your paper?

**Comment by ChandlerGuan** [comment]
Hello @yimingsh, thank you for following our work.
The easiest way to evaluate speedup is to make the evaluation batch size to 1 with  `--per_gpu_eval_batch_size==1` in the launching script. The reason is that transkimmer only masks the pruned tokens during training instead of actually discarding them.

**Comment by yimingsh** [comment]
I have set the bsz=1, but the MAC of Transkimmer is still larger than the BERT. I think the main reason is that the code still masks the token as the training stage.

**Comment by yimingsh** [comment]
Maybe we can remove tokens based on the `skim_mask` when bsz=1? 
BTW, I notice that in the code, you compute the value of remained token and the all tokens. Can the `tokens_remained ratio` also reflect the acceleration of FLOPS? Because I tested on SST2 task follow the hyper-parameters in your paper and got the token_ramained ratio is 0.67. It seems like the 1.58x speedup in your paper?

**Comment by ChandlerGuan** [comment]
> I have set the bsz=1, but the MAC of Transkimmer is still larger than the BERT. I think the main reason is that the code still masks the token as the training stage.

You are right. I forgot to merge the pruning branch when publishing the code base. Sorry for the mistake. I have just pushed a branch with the core implementation. I will further test whether it works fine before merging it to the master branch. You can use this branch at the moment.

> Can the tokens_remained ratio also reflect the acceleration of FLOPS?

We do use the `tokens_remained_ratio` during our development and hyper-parameter tuning. Because evaluation with batch size 1 is rather time-consuming. The `tokens_remained_ratio` is almost identical to speedup.

**Comment by yimingsh** [comment]
Thanks for your reply. I use the branch `feat_skim_eval` and find the speedup is similar as the paper now. 
But I also observed a drop of accuracy, like 92 -> 89, when I use the batch_size=1. Is this normal or I maybe misunderstood something？When the batch size=32, the results is ~92.

**Comment by ChandlerGuan** [comment]
I can reproduce your result. And when `tokens_remained` is small, for example 0.3, the accuracy drop is extremely large 90% -> 30%. It seems like a bug. Looking into it.

**Comment by yimingsh** [comment]
Is it caused by not using Gumbel softmax sampling when bsz=1?

**Comment by yimingsh** [comment]
Hi~ @ChandlerGuan  I found some tiny things may have led to such results: 
(1) When the bsz=1 evaluation, the `skim_loss` isn't accurate because in the code:
``
skim_loss += accumulated_skim_mask/mask.shape[1]
``
The shape of the mask has been reduced by `skim_mask = trunc_with_mask_batched(skim_mask, bool_skim_mask, 1)`. We should use the original shape of skim_mask?
So we can observe the skim_loss is higher (0.87) but tokens_remained is also 0.67 as the bsz>1. But I don't think this could lead an accuracy drop.
(2) I think the ACC drop maybe because of the computation of attention as below?
```
        if attention_mask is not None:
            attention_scores = attention_scores + attention_mask
        # Normalize the attention scores to probabilities.
        attention_probs = nn.Softmax(dim=-1)(attention_scores)
        # mask attention probs during training for skimming
        attention_probs = attention_probs * skim_mask[:, None, None, :]
```
I notice that you mask the attention scores after the softmax; the value of remained token maybe also scaled by other dropped token (skim_mask=0). However, when the bsz=1, this token has been dropped before the self-attention block. Maybe there is a mismatch between the training and the inference? But when I add the skim_mask also before the softmax as the attention_mask as follows, I observe a collapse result: the tokens are all remained. Maybe any training bug? Here is my code.
```
        if attention_mask is not None:
            # Apply the attention mask is (precomputed for all layers in BertModel forward() function)
            attention_scores = attention_scores + attention_mask
        if skim_mask is not None:
            skim_mask = skim_mask[:, None, None, :]
            skim_mask = skim_mask.to(dtype=self.query.weight.dtype)  # fp16 compatibility
            skim_mask = (1.0 - skim_mask) * -10000.0
            attention_scores = attention_scores + skim_mask
        # Normalize the attention scores to probabilities.
        attention_probs = nn.Softmax(dim=-1)(attention_scores)
```

**Comment by ChandlerGuan** [comment]
Hello @yimingsh, thank you for your reply.
I also find that it is a bug in the calculation of the masked attention as you mentioned in (2). The current design masks the attention after softmax leading to a decrease of softmax values in skimming inference. We should mask it before softmax. I think your collapse result is caused by your implementation. Masking it by multiplying a large constant makes the gradient too small. dd44976bc75caf92cf104784ed901044a4a27877 changes the masking of attention by performing it within the softmax calculation. That is, Eq.(4) in the paper should be revised from $SkimAttn(H^i)=Attn(H^i)*M^i=Softmax((Q(H^i)K(H^i))*M^i$ to $SkimAttn(H^i)=MakedSoftmax((Q(H^i)K(H^i), M^i)=\frac{exp((Q(H^i)K(H^i))*M^i+\epsilon}{\sum{exp((Q(H^i)K(H^i))*M^i+\epsilon}}$. I have launched some trivial experiments and it seems the fix won't hurt the performance. Though the hyper-parameters are changed and should be tuned.
Again, thank you for pointing out the bug.

**Comment by yimingsh** [comment]
Hi @ChandlerGuan  I test your code and the performance and speedup is normal now when bsz=1. Maybe there is no bug now :) I also find the `skim_coefficient` need to be larger than the previous to obtain a similar speedup.

**Comment by yimingsh** [comment]
Here is the eval result when skim_coefficient=2, `eval: {'accuracy': 0.9094036697247706}, skim_loss: 0.9994926452636719, tokens_remained: 1.9088189601898193, layers: 0 1.9097338914871216 1 1.9095579385757446 2 1.909165382385254 3 1.9091202020645142 4 1.909033179283142 5 1.9088621139526367 6 1.908818006515503 7 1.9085042476654053 8 1.9084622859954834 9 1.9083331823349 10 1.9082452058792114 11 1.9079915285110474`

**Comment by ChandlerGuan** [comment]
Hello @yimingsh , thank you for your help on this issue. I also find that the skim loss is _harder_ to drop after the fix, or as you suggested, it needs a larger `skim_coefficient`. I tried changing the skim predictor from 2-layer MLP to 3 layers. It makes the optimization of skim loss easier with the price of extra computational overhead. It could be regarded as an architectural hyper-parameter to tune. Though we find that 2-layer is good enough in previous experiments.
I have had a discussion with my supervisor on this problem. In order to fix it, we will re-do the experiments with the new Eq.(4)
and update the data. We are going to revise the paper of both the arxiv and published version (which may take some time).

**Comment by yimingsh** [comment]
Thanks a lot. Looking forward to new results. BTW, I think the current results may still be somewhat convincing.
